### PR TITLE
snap: add yaml to build snap image

### LINKED ISSUE #81: Add yaml to build snap image
Add yaml to build a snap image with Kata Containers inside

### PR DESCRIPTION
**Pull request by devimc**
the yaml file is the recipe to build a snap image
with all Kata Containers components inside.

fixes #81

Signed-off-by: Julio Montes <julio.montes@intel.com>

### CODE REVIEW

File: snap/README.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# Kata Containers Snap
+
+This directory contains the resources needed to build the Kata Containers
+[snap][1] image.
+
+## Initial setup
+
+*Ubuntu Bionic*
```
Comment (jodh-intel) [inline snap/README.md]: Could you either replace with the version number of add it (since the versions numbers are more meaningful than the names imho ;)

File: snap/README.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
+# Kata Containers Snap
```
Comment (jodh-intel) [inline snap/README.md]: A TOC would be useful.

File: snap/README.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+Run next command at the root directory of the packaging repository.
+
+```
+$ snapcraft -d
+```
+
+## Install snap (developer)
+
+```
+$ sudo snap install --classic --dangerous kata-containers_[VERSION]_[ARCH].snap
```
Comment (jodh-intel) [inline snap/README.md]: Could you add a note explaining what `--classic` and particularly what `--dangerous` does? :)

File: snap/README.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+```
+
+Replace `VERSION` with the current version of Kata Containers and `ARCH` with
```
Comment (jodh-intel) [inline snap/README.md]: Couldn't we replace `[ARCH]` with `$(arch)`?

File: snap/README.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+## Configuring Kata Containers ##
+
+
+## Snap vs Packages ##
+
+
+## Limitations
+
+The miniOS image is not included in the snap image due to is not possible to
```
Comment (jodh-intel) [inline snap/README.md]: - Could you add a link to osbuilder and/or https://github.com/kata-containers/documentation/blob/master/architecture.md#root-filesystem-image.

- I'd reword this sentence with something like:
  > The miniOS image is not included in the snap image as it is not possible for QEMU to open a guest RAM backing store on a read-only filesystem.

File: snap/snapcraft.yaml [hunk 6]
```diff
@@ -0,0 +1,253 @@
+name: kata-containers
+version: 1.0.0
```
Comment (jodh-intel) [inline snap/snapcraft.yaml]: Ideally, this file would be `snap/snapcraft.yaml.in` and we've build it to avoid hard-coding all these versions numbers.

I'm also wondering what we can do (if anything) to avoid the script sections from being hard-coded.

Any thoughts @jcvenegas?

Final question - can we attempt to build this snap package under Ubuntu CI systems to ensure we haven't borked it when it's updated?

**Comment by jodh-intel** [comment]
@klynnrif, @intelkevinputnam - ptal.

File: snap/README.md (hunk 4 above)
Comment (devimc) [reply snap/README.md]: nop, for example on x86_64 systems $(arch) prints x86_64, but in snapcraft ARCH is amd64

File: snap/README.md (hunk 5 above)
Comment (devimc) [reply snap/README.md]: will do, thanks

File: snap/README.md (hunk 4 above)
Comment (jodh-intel) [reply snap/README.md]: Ah - ack. We could point users at [`kata-arch.sh --golang`](https://github.com/kata-containers/tests/blob/master/.ci/kata-arch.sh) but it's a pita that `arch(1)` doesn't have these options itself ;(

File: snap/snapcraft.yaml (hunk 6 above)
Comment (devimc) [reply snap/snapcraft.yaml]: > Final question - can we attempt to build this snap package under Ubuntu CI systems to ensure we haven't borked it when it's updated?

yes, we can :)

**Comment by devimc** [comment]
@jodh-intel changes applied, thanks for reviewing this

File: Makefile [hunk 7]
```diff
@@ -19,3 +22,22 @@ test-static-build:
 
 test-packaging-tools:
 	@$(MK_DIR)/build_from_docker.sh
+
+$(SNAPCRAFT_FILE): %: %.in Makefile obs-packaging/versions.txt
```
Comment (jcvenegas) [inline Makefile]: I'd like to remove as much as we could the depedency of this file.  Could you use `versions.yaml` from runtime ?

File: .ci/run.sh [hunk 8]
```diff
@@ -11,8 +11,14 @@ set -o pipefail
...
+source /etc/os-release
 
 # This script will execute packaging tests suite
 # TODO: Add steps needed to build packages
 
 check_kata_kernel_version
+
+echo  "Building snap image"
+if [ "$ID" == ubuntu ];then
+	make snap
```
Comment (jodh-intel) [inline .ci/run.sh]: \o/

File: snap/README.md [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+* [Limitations](#limitations)
+
+# Kata Containers snap image
+
+This directory contains the resources needed to build the Kata Containers
+[snap][1] image.
+
+## Initial setup
+
+*Ubuntu 18 (Bionic Beaver)*
```
Comment (jodh-intel) [inline snap/README.md]: Nit: could you make that "18.04" as "even major number" + "04" minor number == LTS :) Plus when 18.10 comes out, this will be wrong as it won't be Bionic any more ;)

File: snap/README.md [hunk 10]
```diff
@@ -0,0 +1,99 @@
+* [Kata Containers snap image](#kata-containers-snap-image)
```
Comment (jodh-intel) [inline snap/README.md]: Thanks for adding! Minor nit - it looks a lot better I find if you put the TOC just below the title.

File: snap/README.md [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+```sh
+$ make snap
+```
+
+## Install snap (developer)
+
+To install the resulting snap image, snap must be put in [classic mode][3] and the
+security confinement must be disabled (*--classic*). Also since the resulting snap
+has not be signed the verification of signature must be omitted (*--dangerous*).
```
Comment (jodh-intel) [inline snap/README.md]: s/not be/not been/

File: snap/README.md [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+```sh
+$ sudo snap install --classic --dangerous kata-containers_[VERSION]_[ARCH].snap
+```
+
+Replace `VERSION` with the current version of Kata Containers and `ARCH` with
+the system architecture.
+
+## Configuring Kata Containers ##
+
+By default Kata Containers snap image is monted at `/snap/kata-containers` as a
```
Comment (jodh-intel) [inline snap/README.md]: s/mounted/mounted/

File: snap/README.md [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+$ $EDITOR /etc/kata-containers/configuration.toml
+```
+
+## Integration with docker ##
+
+the path to the runtime provided by the Kata Containers snap image is
+`/snap/kata-containers/current/usr/bin/kata-runtime`, this runtime must be added to
+[dockerd][5] via `systemd` or `dockerd` configuration file.
+
+*/etc/systemd/system/docker.service.d/runtime.conf*
```
Comment (jodh-intel) [inline snap/README.md]: Nit: could you make this path (and the one below) fixed-width by using backticks?

File: snap/README.md [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+
+```ini
```
Comment (jodh-intel) [inline snap/README.md]: Cute.

File: snap/README.md (hunk 13 above)
Comment (grahamwhaley) [reply snap/README.md]: I agree, but noted in some of our other docs recently that this was not done. @jodh-intel @klynnrif , shall we agree a standard for this, so at least next time we add or update one we can make them uniform?

File: Makefile (hunk 7 above)
Comment (devimc) [reply Makefile]: mmm but `include` wouldn't work

**Comment by devimc** [comment]
@jcvenegas @jodh-intel changes applied, thanks

**Comment by jodh-intel** [comment]
lgtm

**Comment by sboeuf** [comment]
@grahamwhaley PTAL !

File: Makefile [hunk 15]
```diff
@@ -6,7 +6,17 @@
 #
 
 MK_DIR :=$(shell dirname $(realpath $(lastword $(MAKEFILE_LIST))))
-.PHONY: test test-release-tools
+SED := sed
+YQ := $(shell go env GOPATH)/bin/yq
```
Comment (grahamwhaley) [inline Makefile]: I guess this won't work with [multi-path GOPATHs](https://golang.org/cmd/go/#hdr-GOPATH_environment_variable) - but then I don't think we've ever seen one in use, so I don't see it as a blocker - we have GOPATH usage like that all over the repos...

File: snap/README.md [hunk 16]
```diff
@@ -0,0 +1,99 @@
+# Kata Containers snap image
+
+* [Kata Containers snap image](#kata-containers-snap-image)
```
Comment (grahamwhaley) [inline snap/README.md]: Heh, I think that `Kata Containers snap image` link is redundant, as it just points to the title directly above it. Nuke if you get a chance?

**Comment by jcvenegas** [comment]
lgtm

File: snap/README.md [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+
+By default Kata Containers snap image is mounted at `/snap/kata-containers` as a
+read-only file system, therefore default configuration file can not be edited.
+Fortunately [kata-runtime][4] supports loading a configuration file from another
+path than the default.
+
+```sh
+$ sudo mkdir -p /etc/kata-containers
+$ sudo cp /snap/kata-containers/current/usr/share/defaults/kata-containers/configuration.toml /etc/kata-containers/
+$ $EDITOR /etc/kata-containers/configuration.toml
```
Comment (grahamwhaley) [inline snap/README.md]: Query - I presume all the default paths in the toml file under /snap/kata-containers has paths that point to that path as well - so nothing actually needs to be edited in the copied toml config file for kata to run in default setup here?

File: snap/README.md [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+```
+
+## Integration with docker ##
```
Comment (grahamwhaley) [inline snap/README.md]: Should/can we point to the Ubuntu docker install directions here? - or add a note on how one follows but modifies those and then point at them? https://github.com/kata-containers/documentation/blob/master/install/docker/ubuntu-docker-install.md  Seems like duplication that we have tried to avoid for the other install guides?

File: snap/README.md [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+}
+```
+
+after having added the new runtime, the service must be reloaded and restarted
+
+```
+$ sudo systemctl daemon-reload
+$ sudo systemctl restart docker
+```
+
```
Comment (grahamwhaley) [inline snap/README.md]: And, does the snap also work with k8s - in which case, should we point off to the generic k8s instructions as well? https://github.com/kata-containers/documentation/blob/master/Developer-Guide.md#run-kata-containers-with-kubernetes

File: snap/README.md [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+## Limitations
+
+The [miniOS image][2] is not included in the snap image as it is not possible for
+QEMU to open a guest RAM backing store on a read-only filesystem.
```
Comment (grahamwhaley) [inline snap/README.md]: and.... what are the implications here? - does the snap therefore not work, and does the user have to install the .img file by hand or with a package on the host and point the toml config at it? This feels a bit like it is dangling and not giving the whole story or solution?

File: snap/README.md [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+
+[1]: https://docs.snapcraft.io/snaps/intro
```
Comment (grahamwhaley) [inline snap/README.md]: Nice use of links btw - not seen that for a while!

File: snap/snapcraft.yaml.in [hunk 22]
```diff
@@ -0,0 +1,263 @@
...
+    plugin: go
+    go-importpath: github.com/kata-containers/runtime
+    build-attributes: [no-patchelf]
+    override-build: |
+      # Don't use installed GOROOT
+      unset GOROOT
+      export GOPATH=$(realpath ../go)
+      cd ${GOPATH}/src/github.com/${SNAPCRAFT_PROJECT_NAME}/runtime
+      make \
+        QEMUPATH=/snap/${SNAPCRAFT_PROJECT_NAME}/current/usr/bin/qemu-system-x86_64 \
```
Comment (grahamwhaley) [inline snap/snapcraft.yaml.in]: Just a note for our other arch's - this is hard wired to x86 - I presume there is snap support for other arch's - so we'll need to update/clone this when we get to supporting them via snaps.

File: snap/snapcraft.yaml.in [hunk 23]
```diff
@@ -0,0 +1,263 @@
...
+  runtime:
+    source: http://github.com/kata-containers/runtime
+    source-type: git
+    source-tag: @KATA_RUNTIME_VERSION@
+    after: [go]
+    plugin: go
+    go-importpath: github.com/kata-containers/runtime
+    build-attributes: [no-patchelf]
+    override-build: |
+      # Don't use installed GOROOT
```
Comment (grahamwhaley) [inline snap/snapcraft.yaml.in]: Can you add a note of why we don't use the installed GOROOT pls.

File: snap/snapcraft.yaml.in [hunk 24]
```diff
@@ -0,0 +1,263 @@
...
+      # static build
+      echo "$(${SNAPCRAFT_STAGE}/qemu/scripts/configure-hypervisor.sh -s qemu) \
+        --disable-rbd
+        --prefix=/snap/${SNAPCRAFT_PROJECT_NAME}/current/usr \
+        --datadir=/snap/${SNAPCRAFT_PROJECT_NAME}/current/usr/share \
+        --libexecdir=/snap/${SNAPCRAFT_PROJECT_NAME}/current/usr/libexec/qemu" \
+        | xargs ./configure
+
+      git clone https://github.com/qemu/keycodemapdb ui/keycodemapdb/
+      cd ui/keycodemapdb; git checkout 10739aa; cd ../..
```
Comment (grahamwhaley) [inline snap/snapcraft.yaml.in]: Do we not hold these magic SHAs in our versions.yaml file - feels like maybe we should? /cc @jodh-intel @jcvenegas

Comment (jodh-intel) [reply snap/snapcraft.yaml.in]: Oooh - good catch - I'd missed that ;)

File: snap/README.md (hunk 17 above)
Comment (devimc) [reply snap/README.md]: yes, all is installed at `/snap/kata-containers/`

**Comment by devimc** [comment]
@grahamwhaley thanks for reviewing this, raised #88  to address your code review,
I'll fix snap/snapcraft.yaml.in in other PR
